File: app/src/main/java/com/imin/newprinter/demo/view/LoadingDialog.java
package com.imin.newprinter.demo.view;

import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.LinearInterpolator;
import android.view.animation.RotateAnimation;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;

import com.imin.newprinter.demo.R;

import java.util.ArrayList;
import java.util.List;

/**
 * @Author: hy
 * @date: 2025/4/21
 * @description:
 */
public class LoadingDialog extends android.app.Dialog {

    private TextView textView;

    public enum AnimationType { SPIN, DOTS }

    private ObjectAnimator spinAnimator;
    private List<ValueAnimator> dotAnimators = new ArrayList<>();

    public LoadingDialog(Context context) {
        super(context, R.style.LoadingDialogStyle);
        initialize();
    }

    private void initialize() {
        setContentView(R.layout.dialog_loading);
        textView = findViewById(R.id.tv_text);
        textView.setText(tips);
        getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        setCancelable(false);
    }

    String tips = "";
    public LoadingDialog setText(String text) {

        if (textView != null) {
            tips = text;
            Log.d("PrintDemo_", "WIFI_CONNECT =>" + text);
            textView.setText(text);
        }
        return this;
    }

    public LoadingDialog setAnimationType(AnimationType type) {
        ImageView ivLoading = findViewById(R.id.iv_loading);
        ViewGroup dotContainer = findViewById(R.id.dot_container);

        if (ivLoading == null || dotContainer == null) return this;

        stopAllAnimations();

        switch (type) {
            case SPIN:
                ivLoading.setVisibility(View.VISIBLE);
                dotContainer.setVisibility(View.GONE);
                startSpinAnimation(ivLoading);
                break;
            case DOTS:
                ivLoading.setVisibility(View.GONE);
                dotContainer.setVisibility(View.VISIBLE);
                startDotAnimation(dotContainer);
                break;
        }
        return this;
    }

    private void startSpinAnimation(ImageView view) {
        stopSpinAnimation();

        spinAnimator = ObjectAnimator.ofFloat(view, "rotation", 0f, 360f);
        spinAnimator.setDuration(1000);
        spinAnimator.setRepeatCount(ValueAnimator.INFINITE);
        spinAnimator.setInterpolator(new LinearInterpolator());
        spinAnimator.start();
    }

    private void startDotAnimation(ViewGroup container) {
        stopDotAnimations();

        for (int i = 0; i < container.getChildCount(); i++) {
            View dot = container.getChildAt(i);

            ValueAnimator animator = ValueAnimator.ofFloat(0f, 1f);
            animator.setDuration(400);
            animator.setRepeatCount(ValueAnimator.INFINITE);
            animator.setStartDelay(i * 200L);
            animator.addUpdateListener(animation -> {
                float value = (float) animation.getAnimatedValue();
                float translationY = -20f * (float) Math.sin(value * Math.PI);
                dot.setTranslationY(translationY);
            });

            animator.start();
            dotAnimators.add(animator);
        }
    }

    @Override
    public void dismiss() {
        stopAllAnimations();
        super.dismiss();
    }

    private void stopAllAnimations() {
        stopSpinAnimation();
        stopDotAnimations();
    }

    private void stopSpinAnimation() {
        if (spinAnimator != null && spinAnimator.isRunning()) {
            spinAnimator.cancel();
        }
    }

    private void stopDotAnimations() {
        for (ValueAnimator animator : dotAnimators) {
            if (animator != null && animator.isRunning()) {
                animator.cancel();
            }
        }
        dotAnimators.clear();
    }
}

File: app/src/main/java/com/imin/newprinter/demo/view/LoadingDialog.kt
package com.imin.newprinter.demo.view

import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.app.Dialog
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.LinearInterpolator
import android.view.animation.RotateAnimation
import android.widget.ImageView
import android.widget.TextView
import com.imin.newprinter.demo.R

/**
 * @Author: hy
 * @date: 2025/4/21
 * @description:
 */
// LoadingDialog.kt
class LoadingDialog(context: Context) : Dialog(context, R.style.LoadingDialogStyle) {

    init {
        setContentView(R.layout.dialog_loading)
        window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        setCancelable(false)
    }

    // 设置加载文案
    fun setText(text: String): LoadingDialog {
        findViewById<TextView>(R.id.tv_text)?.text = text
        return this
    }

    // 设置动画类型
    fun setAnimationType(type: AnimationType): LoadingDialog {
        val ivLoading = findViewById<ImageView>(R.id.iv_loading)
        val dotContainer = findViewById<ViewGroup>(R.id.dot_container)

        when(type) {
            AnimationType.SPIN -> {
                ivLoading.visibility = View.VISIBLE
                dotContainer.visibility = View.GONE
                startSpinAnimation(ivLoading)
            }
            AnimationType.DOTS -> {
                ivLoading.visibility = View.GONE
                dotContainer.visibility = View.VISIBLE
                startDotAnimation(dotContainer)
            }
        }
        return this
    }

    // 旋转动画
    private fun startSpinAnimation(view: ImageView) {
        val rotate = RotateAnimation(
            0f, 360f,
            Animation.RELATIVE_TO_SELF, 0.5f,
            Animation.RELATIVE_TO_SELF, 0.5f
        ).apply {
            duration = 1000
            repeatCount = Animation.INFINITE
            interpolator = LinearInterpolator()
        }
        view.startAnimation(rotate)
    }

    // 圆点跳动动画
    private fun startDotAnimation(container: ViewGroup) {
        for (i in 0 until container.childCount) {
            val dot = container.getChildAt(i)
            val anim = ObjectAnimator.ofFloat(dot, "translationY", 0f, -20f, 0f).apply {
                duration = 600
                repeatCount = ValueAnimator.INFINITE
                startDelay = (i * 200).toLong()
            }
            anim.start()
        }
    }

    enum class AnimationType { SPIN, DOTS }
}